ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/Conversions.scala

```diff
@@ -246,27 +239,76 @@ private[state] object Conversions {
-      .addAllDivulgences(encodeDivulgence(blindingInfo.divulgence).asJava)
+      .addAllDivulgences(encodeDivulgence(blindingInfo.divulgence, divulgedContracts).asJava)
       .build
 
-  def decodeBlindingInfo(blindingInfo: DamlTransactionBlindingInfo): BlindingInfo =
-    BlindingInfo(
-      disclosure = blindingInfo.getDisclosuresList.asScala.map { disclosureEntry =>
+  def decodeBlindingInfo(
+      damlTransactionBlindingInfo: DamlTransactionBlindingInfo
+  ): BlindingInfo = {
+    val blindingInfoDivulgence =
+      damlTransactionBlindingInfo.getDivulgencesList.asScala.iterator.map { divulgenceEntry =>
+        val contractId = decodeContractId(divulgenceEntry.getContractId)
+
+        val divulgedTo = divulgenceEntry.getDivulgedToLocalPartiesList.asScala.toSet
+          .map(Ref.Party.assertFromString)
+
+        contractId -> divulgedTo
+      }.toMap
+
+    val blindingInfoDisclosure = damlTransactionBlindingInfo.getDisclosuresList.asScala.map {
+      disclosureEntry =>
         decodeTransactionNodeId(
           disclosureEntry.getNodeId
         ) -> disclosureEntry.getDisclosedToLocalPartiesList.asScala.toSet
           .map(Ref.Party.assertFromString)
-      }.toMap,
-      divulgence = blindingInfo.getDivulgencesList.asScala.map { divulgenceEntry =>
-        decodeContractId(
-          divulgenceEntry.getContractId
-        ) -> divulgenceEntry.getDivulgedToLocalPartiesList.asScala.toSet
-          .map(Ref.Party.assertFromString)
-      }.toMap,
+    }.toMap
+
+    BlindingInfo(
+      disclosure = blindingInfoDisclosure,
+      divulgence = blindingInfoDivulgence,
     )
+  }
+
+  def extractDivulgedContracts(
+      damlTransactionBlindingInfo: DamlTransactionBlindingInfo
+  ): Either[Seq[String], Map[ContractId, Value.ContractInst[VersionedValue[ContractId]]]] = {
+    val divulgences = damlTransactionBlindingInfo.getDivulgencesList.asScala.toVector
+
+    if (divulgences.isEmpty) {
+      Right(Map.empty)
+    } else {
+      val resultAccumulator = Right(
+        Map.newBuilder[ContractId, Value.ContractInst[VersionedValue[ContractId]]]
+      ).withLeft[Seq[String]]
+
+      divulgences
+        .foldLeft(resultAccumulator) {
+          case (Right(contractInstanceIndex), divulgenceEntry) =>
+            if (divulgenceEntry.hasContractInstance) {
+              val contractId = decodeContractId(divulgenceEntry.getContractId)
+              val contractInstance = decodeContractInstance(divulgenceEntry.getContractInstance)
+              Right(contractInstanceIndex += (contractId -> contractInstance))
+            } else {
+              Left(Vector(divulgenceEntry.getContractId))
+            }
+          case (Left(missingContracts), divulgenceEntry: DivulgenceEntry) =>
+            // If populated by an older version of the KV WriteService, contract instances can be missing
+            // Hence, we assume that, if one is missing, all are and return the list of missing ids
```

Comment:
Nit, to make sure that it can't be understood as "f populated by an older version of the KV WriteService, (some) contract instances can be missing":

```suggestion
            // If populated by an older version of the KV WriteService, the contract instances will be missing.
            // Hence, we assume that, if one is missing, all are and return the list of missing ids.
```

Reply:
Done